lib/label.ml

```diff
@@ -0,0 +1,202 @@
+(*
+ * Copyright (c) 2018 Thomas Gazagnaire <thomas@gazagnaire.org>
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+open Compat
+open Result
+
+module Relation = struct
+  type t = Eq | Neq | Le | Lt | Ge | Gt
+
+  let pp ppf = function
+    | Eq -> Fmt.string ppf "="
+    | Neq -> Fmt.string ppf "<>"
+    | Gt -> Fmt.string ppf ">"
+    | Ge -> Fmt.string ppf ">="
+    | Lt -> Fmt.string ppf "<"
+    | Le -> Fmt.string ppf "<="
+
+  let compare = function
+    | Eq -> ( = )
+    | Neq -> ( <> )
+    | Lt -> ( < )
+    | Le -> ( <= )
+    | Gt -> ( > )
+    | Ge -> ( >= )
+
+  let raw_parse_1_char op op_char s =
+    match String.index_opt s op_char with
+    | Some i ->
+      let before = String.sub s 0 i in
+      let after = String.sub s (i+1) (String.length s - i - 1) in
+      (before, Some (op, after))
+    | None -> (s, None)
+
+  let raw_parse_2_chars op op_char_1 op_char_2 s =
+    match String.index_opt s op_char_1 with
+    | Some i
+      when String.length s >= i + 2
+        && Char.equal (String.get s (i+1)) op_char_2 ->
+      let before = String.sub s 0 i in
+      let after = String.sub s (i+2) (String.length s - i - 2) in
+      (before, Some (op, after))
+    | _ -> (s, None)
+
+  let ( ||| ) x y =
+    match x with
+    | _, None -> y
+    | x -> x
+
+  let raw_parse s =
+    raw_parse_2_chars Neq '<' '>' s |||
+    raw_parse_2_chars Ge '>' '=' s |||
+    raw_parse_1_char Gt '>' s |||
+    raw_parse_2_chars Le '<' '=' s |||
+    raw_parse_1_char Lt '<' s |||
+    raw_parse_1_char Eq '=' s
```

Comment:
I think we should use a regex here. This parsing function looks very hard to maintain and is not right:
`dir=a<>b` is interpreted as `"dir=a", Some (Neq, "b")`

Reply:
Done